### ISSUE TITLE
Allow running two modules/services at the same time

### ISSUE THREAD

**Issue by zugaldia** [GoogleCloudPlatform/cloud-code-intellij#1025]
I've created a new project that uses the new Gradle plugin to manage two different modules/services in one AppEngine app.

When I launch either independently they run fine. When I try both at the same time, even if neither is still using the datastore, I get the following error in the second module that has been launched:

```
OperationalError: database is locked
```

I'm using v16.10.4 of the Intellij plugin, v1.0.0 of the Gradle plugin, and gcloud components is up to date. I've tried setting a separate datastore path for each project with `-Ddatastore.backing_store=/tmp/gae_<modulename>.db` without success.

Possibly related to https://github.com/GoogleCloudPlatform/google-cloud-intellij/issues/941.

**Comment by patflynn** [comment]
Thank you for your report!

We are aware the multi-module development in our plugin falls short of expectations.  The issue may be  broader issue with the dev appserver included with the Cloud SDK.

If you could provide a sample project that we could reproduce that would be super helpful. 

Thanks!

**Comment by zugaldia** [comment]
> If you could provide a sample project that we could reproduce that would be super helpful.

Will do.

Meanwhile, is there any available workaround? Downgrading, running from out side IDE?

**Comment by patflynn** [comment]
The only work-around that I know of is using the App Engine SDK based tooling. 

e.g. Disable or uninstall our plugins, and re-able the Jetbrains App Engine integration.

@ludoch 
@arouzrokh
@rdayal

**Comment by patflynn** [comment]
@zugaldia actually I think I may be getting my wirese crossed here.  Are you running the dev appserver from Gradle rather than from the IJ run config?

**Comment by patflynn** [comment]
Which Gradle plugin are you using?

**Comment by patflynn** [comment]
@loosebazooka

**Comment by zugaldia** [comment]
Using https://github.com/GoogleCloudPlatform/gradle-appengine-plugin for Gradle.

I'm running the devserver from the Intellij plugin.

**Comment by zugaldia** [comment]
To be specific: `classpath 'com.google.cloud.tools:appengine-gradle-plugin:1.0.0'` (got it from Maven, not from GitHub).

**Comment by chanseokoh** [comment]
One workaround could be to run `dev_appserver.py` on a terminal outside the IDE. It has the following option:

```
  --storage_path PATH   path to the data (datastore, blobstore, etc.)
                        associated with the application. (default: None)
```

This will certainly use the given path (directory) to store various data including the datastore, so if you use different paths, you should not see the error `OperationalError: database is locked`.

**Comment by patflynn** [comment]
@zugaldia I'm surprised you are able to run multiple modules from the intellij plugin, since we only allow you to select one artifact to deploy.  Or are you running multiple development server run configurations?  If that's the case that would explain your issue.  @ludoch I don't think that configuration can be made to work can it?  

The only viable way I can think of for running multiple modules is by running the dev appserver from Gradle. @loosebazooka

**Comment by zugaldia** [comment]
@patflynn I should've clarified, I created a different configuration for each module selecting the appropriate artifact and then I'm running multiple configurations at the same time (the second one fails).

**Comment by loosebazooka** [comment]
This is probably possible to do in gradle directly, if you chose to run your devappserver from gradle.
What you need to do is configure the parameter in the dev appserver manually to launch both modules.

Possibly in a third build.gradle (or in one of the modules)
```
appengine {
  run {
    appYamls = ["module1/build/exploded-app", "module2/build/exploded-app"]
  }
}

appengineRun.dependsOn "module1:assemble:", ":module2:assemble"
```
and it should launch both modules in one devappserver

**Comment by zugaldia** [comment]
@loosebazooka Thank you, for some reason that didn't work for me, I'm getting the following error about  the `appYamls` setting when running Gradle:

```
* What went wrong:
Execution failed for task ':appweb:appengineRun'.
> java.lang.String cannot be cast to java.io.File
```

Meanwhile, I've confirmed that for the same project, calling `dev_appserver.py` directly works:

```
dev_appserver.py \
	module1/build/exploded-app \
	module2/build/exploded-app \
```

It also works if you set a different port and `storagePath` for each module in Gradle with something like:

```
appengine {
    run {
        storagePath = "/tmp/gae_module1.db"
        port = 8080
        adminPort = 8000
    }
}
```

Coming back to the IntelliJ  plugin, I think we could have this running too if either of these options get implemented:

1. The possibility to choose multiple artifacts in the App Engine Configuration (preferable).
2. The possibility to pass parameters to `dev_appserver.py` so that we can set a custom `storagePath`.

**Comment by patflynn** [comment]
Hi Antonio,

Are you OK with the two services using different storage?  That wouldn't
happen when you were actually running in a real project.

On Fri, Oct 14, 2016 at 11:30 AM, Antonio Zugaldia <notifications@github.com

**Comment by patflynn** [comment]
@zugaldia also do you have a full stacktrace from that gradle error?

**Comment by zugaldia** [comment]
> Are you OK with the two services using different storage?

This isn't a deal-breaker for this particular project (for now) because of the way it's archictected, however if I were given the option, I'd prefer seeing the plugin letting me launch several artifacts at once (sharing the datastore) rather than having to specify different storage paths.

> also do you have a full stacktrace from that gradle error?

Here you go, re-ran Gradle with the `--debug option`, I'm afraid it doesn't seem very useful:

```
10:36:28.271 [DEBUG] [org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter] Determining if task ':modulename:war' is up-to-date
10:36:28.279 [INFO] [org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter] Skipping task ':modulename:war' as it is up-to-date (took 0.007 secs).
10:36:28.279 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Finished executing task ':modulename:war'
10:36:28.279 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :modulename:war UP-TO-DATE
10:36:28.279 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :modulename:war (Thread[Daemon worker,5,main]) completed. Took 0.025 secs.
10:36:28.279 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :modulename:explodeWar (Thread[Daemon worker,5,main]) started.
10:36:28.279 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :modulename:explodeWar
10:36:28.279 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Starting to execute task ':modulename:explodeWar'
10:36:28.279 [DEBUG] [org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter] Determining if task ':modulename:explodeWar' is up-to-date
10:36:28.281 [INFO] [org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter] Skipping task ':modulename:explodeWar' as it is up-to-date (took 0.001 secs).
10:36:28.282 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Finished executing task ':modulename:explodeWar'
10:36:28.282 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :modulename:explodeWar UP-TO-DATE
10:36:28.292 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :modulename:explodeWar (Thread[Daemon worker,5,main]) completed. Took 0.013 secs.
10:36:28.292 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :modulename:assemble (Thread[Daemon worker,5,main]) started.
10:36:28.292 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :modulename:assemble
10:36:28.292 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Starting to execute task ':modulename:assemble'
10:36:28.292 [INFO] [org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter] Skipping task ':modulename:assemble' as it has no actions.
10:36:28.293 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Finished executing task ':modulename:assemble'
10:36:28.293 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :modulename:assemble UP-TO-DATE
10:36:28.293 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :modulename:assemble (Thread[Daemon worker,5,main]) completed. Took 0.001 secs.
10:36:28.293 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :modulename:appengineRun (Thread[Daemon worker,5,main]) started.
10:36:28.293 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :modulename:appengineRun
10:36:28.293 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Starting to execute task ':modulename:appengineRun'
10:36:28.293 [DEBUG] [org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter] Determining if task ':modulename:appengineRun' is up-to-date
10:36:28.293 [INFO] [org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter] Executing task ':modulename:appengineRun' (up-to-date check took 0.0 secs) due to:
  Task has not declared any outputs.
10:36:28.294 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter] Executing actions for task ':modulename:appengineRun'.
10:36:28.295 [DEBUG] [org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter] Finished executing task ':modulename:appengineRun'
10:36:28.295 [LIFECYCLE] [class org.gradle.TaskExecutionLogger] :modulename:appengineRun FAILED
10:36:28.295 [INFO] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] :modulename:appengineRun (Thread[Daemon worker,5,main]) completed. Took 0.002 secs.
10:36:28.295 [DEBUG] [org.gradle.execution.taskgraph.AbstractTaskPlanExecutor] Task worker [Thread[Daemon worker,5,main]] finished, busy: 0.196 secs, idle: 0.023 secs
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] 
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] FAILURE: Build failed with an exception.
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] 
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] * What went wrong:
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] Execution failed for task ':modulename:appengineRun'.
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] > java.lang.String cannot be cast to java.io.File
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] 
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] * Try:
10:36:28.296 [ERROR] [org.gradle.BuildExceptionReporter] Run with --stacktrace option to get the stack trace. 
10:36:28.296 [LIFECYCLE] [org.gradle.BuildResultLogger] 
10:36:28.296 [LIFECYCLE] [org.gradle.BuildResultLogger] BUILD FAILED
10:36:28.296 [LIFECYCLE] [org.gradle.BuildResultLogger] 
10:36:28.296 [LIFECYCLE] [org.gradle.BuildResultLogger] Total time: 1.893 secs
```

(I replaced the actual module name with `modulename`).

**Comment by patflynn** [comment]
Thanks!  Could you try running with '-- stacktrace' to see if it outputs
more?

**Comment by loosebazooka** [comment]
Sorry it looks like the parameter is a file. You have to say [new
File("/path/to/project1/build/exploded-app"), new
File("/path/to/project2/build/exploded-app")]

**Comment by zugaldia** [comment]
@loosebazooka thank you, that did the trick. Also, I needed to make sure to run `./gradlew :module1:appengineRun` instead of simply `./gradlew appengineRun` for that config to apply (where `module1` is the module where I added the Gradle rules). I chose the `default` module but I imagine that any module would work.

For those following at home, the original issue with the IntelliJ plugin is still present as explained in https://github.com/GoogleCloudPlatform/google-cloud-intellij/issues/1025#issuecomment-253835000.

Thank you all for your help so far finding a good workaround.
